In [1]:
# IMPORT THE PACKAGES
import pandas as pd
import numpy as np
import sklearn
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn import metrics
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import cross_val_score
from imblearn.combine import SMOTEENN
from sklearn.neighbors import KNeighborsClassifier
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler, LabelEncoder

In [2]:
# IMORT THE DATA
column = ['duration',' protocol_type',' service',' flag', ' src_bytes',' dst_bytes', ' land', ' wrong_fragment',
                            ' urgent',                         ' hot',
                 ' num_failed_logins',                   ' logged_in',
                   ' num_compromised',                  ' root_shell',
                      ' su_attempted',                    ' num_root',
                ' num_file_creations',                  ' num_shells',
                  ' num_access_files',           ' num_outbound_cmds',
                     ' is_host_login',              ' is_guest_login',
                             ' count',                   ' srv_count',
                       ' serror_rate',              ' srv_error_rate',
                       ' rerror_rate',             ' srv_rerror_rate',
                     ' same_srv_rate',               ' diff_srv_rate',
                ' srv_diff_host_rate',              ' dst_host_count',
                ' dst_host_srv_count',      ' dst_host_same_srv_rate',
            ' dst_host_diff_srv_rate', ' dst_host_same_src_port_rate',
       ' dst_host_srv_diff_host_rate',        ' dst_host_serror_rate',
          ' dst_host_srv_serror_rate',        ' dst_host_rerror_rate',
                ' dst_host_srv_rerror_rate']

In [3]:
file=['Data_of_Attack_Back.csv','Data_of_Attack_Back_BufferOverflow.csv',
       'Data_of_Attack_Back_FTPWrite.csv','Data_of_Attack_Back_GuessPassword.csv',
       'Data_of_Attack_Back_NMap.csv','Data_of_Attack_Back_Neptune.csv',
       'Data_of_Attack_Back_Normal.csv','Data_of_Attack_Back_PortSweep.csv',
       'Data_of_Attack_Back_RootKit.csv','Data_of_Attack_Back_Satan.csv',
       'Data_of_Attack_Back_Smurf.csv'
      ]

In [4]:
label=['back', 'buff', 'ftp', 'guess', 'nep', 'nmap', 'normal', 'port', 'root', 'satan', 'smurf']

In [5]:
data1=[]
for f,l in zip(file,label):
    if f=='Data_of_Attack_Back_FTPWrite.csv':
        a=pd.read_csv(f,header=None,names=column)
    else:
        a = pd.read_csv(f)

    a['Label'] = l
    data1.append(a)

# Combine into a single DataFrame
data = pd.concat(data1, ignore_index=True)

In [6]:
data

duration   protocol_type   service   flag   src_bytes   dst_bytes  \
0            0.0            0.00      0.00    0.0     0.54540     0.08314   
1            0.0            0.00      0.00    0.0     0.54540     0.08314   
2            0.0            0.00      0.00    0.0     0.54540     0.08314   
3            0.0            0.00      0.00    0.0     0.54540     0.08314   
4            0.0            0.00      0.00    0.0     0.54540     0.08314   
...          ...             ...       ...    ...         ...         ...   
817546       0.0            0.02      0.09    0.0     0.01032     0.00000   
817547       0.0            0.02      0.09    0.0     0.01032     0.00000   
817548       0.0            0.02      0.09    0.0     0.01032     0.00000   
817549       0.0            0.02      0.09    0.0     0.01032     0.00000   
817550       0.0            0.01      0.12    0.0     0.00028     0.00000   

         land   wrong_fragment   urgent   hot  ...   dst_host_srv_count  \
0           0              0.0      0.0   0.2  ...                0.001   
1           0              0.0      0.0   0.2  ...                0.002   
2           0              0.0      0.0   0.2  ...                0.003   
3           0              0.0      0.0   0.2  ...                0.004   
4           0              0.0      0.0   0.2  ...                0.005   
...       ...              ...      ...   ...  ...                  ...   
817546      0              0.0      0.0   0.0  ...                0.251   
817547      0              0.0      0.0   0.0  ...                0.252   
817548      0              0.0      0.0   0.0  ...                0.253   
817549      0              0.0      0.0   0.0  ...                0.254   
817550      0              0.3      0.0   0.0  ...                0.001   

         dst_host_same_srv_rate   dst_host_diff_srv_rate  \
0                         0.100                    0.000   
1                         0.100                    0.000   
2                         0.100                    0.000   
3                         0.100                    0.000   
4                         0.100                    0.000   
...                         ...                      ...   
817546                    0.098                    0.001   
817547                    0.099                    0.001   
817548                    0.099                    0.001   
817549                    0.100                    0.001   
817550                    0.000                    0.001   

         dst_host_same_src_port_rate   dst_host_srv_diff_host_rate  \
0                              0.100                           0.0   
1                              0.050                           0.0   
2                              0.033                           0.0   
3                              0.025                           0.0   
4                              0.020                           0.0   
...                              ...                           ...   
817546                         0.098                           0.0   
817547                         0.099                           0.0   
817548                         0.099                           0.0   
817549                         0.100                           0.0   
817550                         0.000                           0.0   

         dst_host_serror_rate   dst_host_srv_serror_rate  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         0.0                        0.0   
4                         0.0                        0.0   
...                       ...                        ...   
817546                    0.0                        0.0   
817547                    0.0                        0.0   
817548                    0.0                        0.0   
817549              

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 817551 entries, 0 to 817550
Data columns (total 42 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   duration                      817551 non-null  float64
 1    protocol_type                817551 non-null  float64
 2    service                      817551 non-null  float64
 3    flag                         817551 non-null  float64
 4    src_bytes                    817551 non-null  float64
 5    dst_bytes                    817551 non-null  float64
 6    land                         817551 non-null  int64  
 7    wrong_fragment               817551 non-null  float64
 8    urgent                       817551 non-null  float64
 9    hot                          817551 non-null  float64
 10   num_failed_logins            817551 non-null  float64
 11   logged_in                    817551 non-null  float64
 12   num_compromised              817551 non-nul

In [8]:
# EXPLORATORY DATA ANALYSIS (EDA)
# HANDLE MISSING VALUES 
data.isna().sum()

duration                        0
 protocol_type                  0
 service                        0
 flag                           0
 src_bytes                      0
 dst_bytes                      0
 land                           0
 wrong_fragment                 0
 urgent                         0
 hot                            0
 num_failed_logins              0
 logged_in                      0
 num_compromised                0
 root_shell                     0
 su_attempted                   0
 num_root                       0
 num_file_creations             0
 num_shells                     0
 num_access_files               0
 num_outbound_cmds              0
 is_host_login                  0
 is_guest_login                 0
 count                          0
 srv_count                      0
 serror_rate                    0
 srv_error_rate                 0
 rerror_rate                    0
 srv_rerror_rate                0
 same_srv_rate                  0
 diff_srv_rate

In [9]:
# there are no null values in the data

In [10]:
data.describe()

duration   protocol_type        service           flag  \
count  817551.000000   817551.000000  817551.000000  817551.000000   
mean        0.000877        0.000570       0.052909       0.020160   
std         0.026636        0.002596       0.096389       0.030959   
min         0.000000        0.000000       0.000000       0.000000   
25%         0.000000        0.000000       0.000000       0.000000   
50%         0.000000        0.000000       0.000000       0.000000   
75%         0.000000        0.000000       0.120000       0.070000   
max         1.000000        0.020000       0.670000       0.100000   

           src_bytes      dst_bytes      land   wrong_fragment         urgent  \
count  817551.000000  817551.000000  817551.0     8.175510e+05  817551.000000   
mean        0.004961       0.024862       0.0     3.669496e-07       0.000003   
std         0.033235       0.066635       0.0     3.317904e-04       0.000805   
min         0.000000       0.000000       0.0     0.000000e+00       0.000000   
25%         0.000000       0.000000       0.0     0.000000e+00       0.000000   
50%         0.002170       0.003670       0.0     0.000000e+00       0.000000   
75%         0.002990       0.019690       0.0     0.000000e+00       0.000000   
max         1.000000       1.000000       0.0     3.000000e-01       0.500000   

                 hot  ...   dst_host_count   dst_host_srv_count  \
count  817551.000000  ...    817551.000000        817551.000000   
mean        0.002456  ...         0.166868             0.161651   
std         0.041470  ...         0.102058             0.111642   
min         0.000000  ...         0.000000             0.000000   
25%         0.000000  ...         0.054000             0.017000   
50%         0.000000  ...         0.255000             0.255000   
75%         0.000000  ...         0.255000             0.255000   
max         1.000000  ...         0.255000             0.255000   

        dst_host_same_srv_rate   dst_host_diff_srv_rate  \
count            817551.000000            817551.000000   
mean                  0.066795                 0.003380   
std                   0.043421                 0.008557   
min                   0.000000                 0.000000   
25%                   0.007000                 0.000000   
50%                   0.100000                 0.000000   
75%                   0.100000                 0.006000   
max                   0.100000                 0.100000   

        dst_host_same_src_port_rate   dst_host_srv_diff_host_rate  \
count                 817551.000000                 817551.000000   
mean                       0.006148                      0.001683   
std                        0.017460                      0.003708   
min                        0.000000                      0.000000   
25%                        0.000000                      0.000000   
50%                        0.000000                      0.000000   
75%                        0.003000                      0.002000   
max                        0.100000                      0.100000   

        dst_host_serror_rate   dst_host_srv_serror_rate  \
count          817551.000000              817551.000000   
mean                0.024865                   0.024852   
std                 0.043062                   0.043151   
min                 0.000000                   0.000000   
25%                 0.000000                   0.000000   
50%                 0.000000                   0.000000   
75%                 0.014000                   0.003000   
max                 0.100000                   0.100000   

        dst_host_rerror_rate   dst_host_srv_rerror_rate  
count          817551.000000              817551.000000  
mean                0.005372                   0.005279  
std                 0.021920                   0.022070  
min                 0.000000                   0.000000  
25%                 0.000000                   0.000000  
50%          

In [11]:
data.Label.value_counts()

Label
normal    576710
nmap      227228
satan       5019
smurf       3007
port        2964
nep         1554
back         968
guess         53
buff          30
root          10
ftp            8
Name: count, dtype: int64

In [12]:
# LOOKING FOR ROWS WHERE LABEL = 'ftp'
data[data.Label=='ftp']

duration   protocol_type   service   flag   src_bytes   dst_bytes  \
998     0.0026             0.0      0.07    0.0     0.00116     0.00451   
999     0.0134             0.0      0.34    0.0     0.00100     0.39445   
1000    0.0000             0.0      0.14    0.0     0.00613     0.00000   
1001    0.0000             0.0      0.14    0.0     0.00000     0.00005   
1002    0.0032             0.0      0.07    0.0     0.00104     0.00449   
1003    0.0067             0.0      0.34    0.0     0.00157     0.02703   
1004    0.0000             0.0      0.14    0.0     0.00676     0.00000   
1005    0.0000             0.0      0.14    0.0     0.00000     0.00005   

       land   wrong_fragment   urgent   hot  ...   dst_host_srv_count  \
998       0              0.0      0.0   0.2  ...                0.001   
999       0              0.0      0.2   0.0  ...                0.001   
1000      0              0.0      0.0   0.0  ...                0.084   
1001      0              0.0      0.0   0.0  ...                0.085   
1002      0              0.0      0.0   0.2  ...                0.001   
1003      0              0.0      0.1   0.0  ...                0.001   
1004      0              0.0      0.0   0.0  ...                0.004   
1005      0              0.0      0.0   0.0  ...                0.005   

       dst_host_same_srv_rate   dst_host_diff_srv_rate  \
998                      0.10                      0.0   
999                      0.05                      0.1   
1000                     0.10                      0.0   
1001                     0.10                      0.0   
1002                     0.10                      0.0   
1003                     0.05                      0.1   
1004                     0.10                      0.0   
1005                     0.10                      0.0   

       dst_host_same_src_port_rate   dst_host_srv_diff_host_rate  \
998                           0.10                         0.000   
999                           0.05                         0.000   
1000                          0.10                         0.002   
1001                          0.10                         0.002   
1002                          0.10                         0.000   
1003                          0.05                         0.000   
1004                          0.10                         0.050   
1005                          0.10                         0.040   

       dst_host_serror_rate   dst_host_srv_serror_rate   dst_host_rerror_rate  \
998                     0.0                        0.0                    0.0   
999                     0.0                        0.0                    0.0   
1000                    0.0                        0.0                    0.0   
1001                    0.0                        0.0                    0.0   
1002                    0.0                        0.0                    0.0   
1003                    0.0                        0.0                    0.0   
1004                    0.0                        0.0                    0.0   
1005                    0.0                        0.0                    0.0   

       dst_host_srv_rerror_rate  Label  
998                         0.0    ftp  
999                         0.0    ftp  
1000                        0.0    ftp  
1001                        0.0    ftp  
1002                        0.0    ftp  
1003                        0.0    ftp  
1004                        0.0    ftp  
1005                        0.0    ftp  

[8 rows x 42 columns]

#### Data Preparation:
You are required to append all the files and create new column called attack based on the name of 
attack. While you are appending the files, you can take resampling of data based on the number of 
attacks.
## For Binomial classification, you can create attack variable with attack vs. normal


In [13]:
# TO PERFORM BINOMIAL CLASSIFICATION I AM USING DESCISSION TREE CLASSIFIER

In [14]:
data['Attack']=data['Label'].apply(lambda x: 0 if x=='normal' else 1)

In [15]:
data.head()

duration   protocol_type   service   flag   src_bytes   dst_bytes   land  \
0       0.0             0.0       0.0    0.0      0.5454     0.08314      0   
1       0.0             0.0       0.0    0.0      0.5454     0.08314      0   
2       0.0             0.0       0.0    0.0      0.5454     0.08314      0   
3       0.0             0.0       0.0    0.0      0.5454     0.08314      0   
4       0.0             0.0       0.0    0.0      0.5454     0.08314      0   

    wrong_fragment   urgent   hot  ...   dst_host_same_srv_rate  \
0              0.0      0.0   0.2  ...                      0.1   
1              0.0      0.0   0.2  ...                      0.1   
2              0.0      0.0   0.2  ...                      0.1   
3              0.0      0.0   0.2  ...                      0.1   
4              0.0      0.0   0.2  ...                      0.1   

    dst_host_diff_srv_rate   dst_host_same_src_port_rate  \
0                      0.0                         0.100   
1                      0.0                         0.050   
2                      0.0                         0.033   
3                      0.0                         0.025   
4                      0.0                         0.020   

    dst_host_srv_diff_host_rate   dst_host_serror_rate  \
0                           0.0                    0.0   
1                           0.0                    0.0   
2                           0.0                    0.0   
3                           0.0                    0.0   
4                           0.0                    0.0   

    dst_host_srv_serror_rate   dst_host_rerror_rate  \
0                        0.0                    0.0   
1                        0.0                    0.0   
2                        0.0                    0.0   
3                        0.0                    0.0   
4                        0.0                    0.0   

    dst_host_srv_rerror_rate  Label  Attack  
0                        0.0   back       1  
1                        0.0   back       1  
2                        0.0   back       1  
3                        0.0   back       1  
4                        0.0   back       1  

[5 rows x 43 columns]

In [16]:
#Prepare the Data

# Features (drop the 'Label' and 'Binary_Label' columns)
X = data.drop(['Label', 'Attack'], axis=1)

# Binary target variable
y = data['Attack']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=5)


In [17]:
# FOR HANDLING IMBALANCED DATA
smote = SMOTE()
X_train_sm, y_train_sm = smote.fit_resample(X_train, y_train)

In [18]:
scaler = StandardScaler()
X_train_sm = scaler.fit_transform(X_train_sm)
X_test = scaler.transform(X_test)

In [19]:
dt_classifier = DecisionTreeClassifier(random_state=5, max_depth=10, min_samples_split=20, min_samples_leaf=10)

In [20]:
dt_classifier

DecisionTreeClassifier(max_depth=10, min_samples_leaf=10, min_samples_split=20,
                       random_state=5)

In [21]:
cv_scores = cross_val_score(dt_classifier, X, y, cv=5) 

In [22]:
cv_scores

array([0.99198219, 0.99995719, 0.99998165, 0.99862394, 0.78804966])

In [23]:
cv_scores.mean()

0.9557189260813518

In [24]:
dt_classifier.fit(X_train,y_train)

DecisionTreeClassifier(max_depth=10, min_samples_leaf=10, min_samples_split=20,
                       random_state=5)

In [25]:
# Predict on the test data
y_pred = dt_classifier.predict(X_test)

C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


In [26]:
accuracy = accuracy_score(y_test, y_pred)

In [27]:
accuracy

0.9602187013283537

In [28]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.94      0.97    172796
           1       0.88      1.00      0.94     72470

    accuracy                           0.96    245266
   macro avg       0.94      0.97      0.95    245266
weighted avg       0.96      0.96      0.96    245266



In [29]:
print(confusion_matrix(y_test, y_pred))

[[163162   9634]
 [   123  72347]]


The classification report you provided contains precision, recall, f1-score, and support values for two classes normal activities and attack
1. Precision:
Class 0 (normal activities): 1.00 — This means that almost all predictions for class 0 were correct.
Class 1 (attacks): 0.88 — Out of all instances predicted as attacks, 88% were correct.
Precision measures how many of the predicted positives are actual positives.

2. Recall:
Class 0 (normal activities): 0.94 — 94% of the actual normal activities were correctly identified.
Class 1 (attacks): 1.00 — All actual attack instances were identified correctly.
Recall shows how well the model identifies actual positive cases.

3. F1-Score:
Class 0: 0.97
Class 1: 0.94
The F1-score balances precision and recall, providing a better measure when classes are imbalanced.

4. Accuracy: 0.96
The model correctly predicted 96% of the total instances.
5. Macro and Weighted Averages:
Macro Avg: 0.94 precision, 0.97 recall, and 0.95 F1-score — This is the unweighted average across the classes.
Weighted Avg: 0.96 precision, recall, and F1-score — This takes class imbalances into account.
6. Confusion Matrix:
True Positives (Class 0): 163,162 (correctly classified as normal activities).
False Positives (Class 0): 9,634 (normal activities misclassified as attacks).
True Positives (Class 1): 72,347 (correctly classified as attacks).
False Negatives (Class 1): 123 (attacks misclassified as normal activities).
Key Observations:
Class 0 (normal activities): High precision but slightly lower recall (0.94), meaning some normal activities were misclassified as attacks (false positives).
Class 1 (attacks): High recall (1.00), but precision is slightly lower (0.88), indicating some false positives (normal activities classified as attacks).


Conclusion:
The model has good overall performance, but there is a trade-off between precision and recall for class 1 (attacks). While it identifies most attacks correctly, it occasionally misclassifies normal activities as attacks. The overall accuracy is strong, but the imbalance between precision and recall for class 1 suggests potential room for improvement in reducing false positives for attacks.




### For Multinomial classification, 
you can create attack variable with normal vs. Back vs. Buffer Over 
flow vs. FTP Write vs. Guess Password vs. Neptune vs. N-Map vs. Port Sweep vs. Root Kit vs. 
Satan vs. Smurf

In [30]:
# TO PERFORM MULTINOMIAL CLASSIFICATION I AM USING KNN CLASSIFIER

In [31]:
X = data.drop(['Label', 'Attack'], axis=1)
y = data['Label']
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=5)

In [32]:
# Initialize the KNN Classifier with k=5 
knn_classifier = KNeighborsClassifier(n_neighbors=5)
knn_classifier

KNeighborsClassifier()

In [33]:
# Fit the model on the training data
knn_classifier.fit(X_train, y_train)

KNeighborsClassifier()

In [34]:
# Make predictions on the test set
y_pred = knn_classifier.predict(X_test)

In [35]:
accuracy = accuracy_score(y_test, y_pred)

In [40]:
print('accuracy',accuracy)

accuracy 0.9996901323461058


In [36]:
print(classification_report(y_test, y_pred))

C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


              precision    recall  f1-score   support

        back       1.00      1.00      1.00       304
        buff       1.00      1.00      1.00         1
         ftp       0.00      0.00      0.00         1
       guess       1.00      1.00      1.00        11
         nep       0.98      0.99      0.99       442
        nmap       1.00      1.00      1.00     68501
      normal       1.00      1.00      1.00    172796
        port       0.99      0.98      0.99       844
        root       0.00      0.00      0.00         3
       satan       0.99      0.98      0.99      1502
       smurf       1.00      0.99      1.00       861

    accuracy                           1.00    245266
   macro avg       0.81      0.81      0.81    245266
weighted avg       1.00      1.00      1.00    245266



C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [37]:
print(confusion_matrix(y_test, y_pred))

[[   303      0      0      0      0      0      1      0      0      0
       0]
 [     0      1      0      0      0      0      0      0      0      0
       0]
 [     0      0      0      0      0      0      1      0      0      0
       0]
 [     0      0      0     11      0      0      0      0      0      0
       0]
 [     0      0      0      0    437      0      4      0      0      1
       0]
 [     0      0      0      0      0  68501      0      0      0      0
       0]
 [     1      0      0      0      5      0 172779      3      0      8
       0]
 [     0      0      0      0      2      1     13    827      0      1
       0]
 [     0      0      0      0      0      0      3      0      0      0
       0]
 [     0      0      0      0      0      0     24      2      0   1476
       0]
 [     0      0      0      0      1      0      4      1      0      0
     855]]


1. Accuracy:
The overall accuracy is 0.9997 (99.97%), which indicates that the model correctly classified almost all samples.
2. Class-wise Performance:
Classes like "back," "buff," "guess," "nmap," "normal" have perfect precision, recall, and F1-scores of 1.00, meaning they are perfectly classified.
Classes like "ftp" and "root" have 0 precision, recall, and F1-scores, meaning the model failed to correctly classify these samples.
Classes like "nep," "port," and "satan" have very high performance, but there are a few misclassifications.
3. Macro and Weighted Averages:
Macro Average: Precision, recall, and F1-score are 0.81. This lower score reflects that some classes with fewer instances (e.g., "ftp," "root") are not well-predicted.
Weighted Average: Precision, recall, and F1-score are all 1.00, indicating that the model performs exceptionally well overall, especially considering the majority class ("normal").
4. Confusion Matrix:
The confusion matrix gives a detailed view of correct and incorrect predictions:

Perfect Predictions: For classes like back, buff, guess, nmap, normal, satan, smurf, almost all instances are predicted correctly.
Misclassifications:
Class 0 (normal activities): 5 normal activities misclassified as "nep," 13 misclassified as "port," and 24 misclassified as "satan."
Class 1 (attacks): Small misclassifications, like 4 "nep" instances classified as "normal" and 4 "port" instances misclassified as "normal."
Key Observations:
The model shows outstanding performance in most cases, especially for large classes such as normal and nmap.
However, there is room for improvement for smaller classes, such as ftp and root, where the model fails to make any correct predictions.
This suggests that while the model handles the majority classes very well, it struggles with some minority classes, which could be due to class imbalance or insufficient data for those categories.






